trinity/main.py

```diff
@@ -206,27 +242,62 @@ def trinity_boot(args: Namespace,
     networking_process.start()
     logger.info("Started networking process (pid=%d)", networking_process.pid)
 
+    main_endpoint.subscribe(
+        ShutdownRequest,
+        lambda ev: kill_trinity_gracefully(
+            logger,
+            database_server_process,
+            networking_process,
+            plugin_manager,
+            event_bus
+        )
+    )
+
+    plugin_manager.prepare(args, chain_config, extra_kwargs)
+    plugin_manager.broadcast(TrinityStartupEvent(
+        args,
+        chain_config
+    ))
     try:
-        networking_process.join()
+        loop = asyncio.get_event_loop()
+        loop.run_forever()
+        loop.close()
     except KeyboardInterrupt:
-        # When a user hits Ctrl+C in the terminal, the SIGINT is sent to all processes in the
-        # foreground *process group*, so both our networking and database processes will terminate
-        # at the same time and not sequentially as we'd like. That shouldn't be a problem but if
-        # we keep getting unhandled BrokenPipeErrors/ConnectionResetErrors like reported in
-        # https://github.com/ethereum/py-evm/issues/827, we might want to change the networking
-        # process' signal handler to wait until the DB process has terminated before doing its
-        # thing.
-        # Notice that we still need the kill_process_gracefully() calls here, for when the user
-        # simply uses 'kill' to send a signal to the main process, but also because they will
-        # perform a non-gracefull shutdown if the process takes too long to terminate.
-        logger.info('Keyboard Interrupt: Stopping')
-        kill_process_gracefully(database_server_process, logger)
-        logger.info('DB server process (pid=%d) terminated', database_server_process.pid)
-        # XXX: This short sleep here seems to avoid us hitting a deadlock when attempting to
-        # join() the networking subprocess: https://github.com/ethereum/py-evm/issues/940
-        time.sleep(0.2)
-        kill_process_gracefully(networking_process, logger)
-        logger.info('Networking process (pid=%d) terminated', networking_process.pid)
+        kill_trinity_gracefully(
+            logger,
+            database_server_process,
+            networking_process,
+            plugin_manager,
+            event_bus
+        )
+
+
+def kill_trinity_gracefully(logger: logging.Logger,
```

Comment:
I wonder if it's worth making a named tuple (e.g `BootArtifacts`) to tighten the function signature